<launch>
  <!-- Launch controller -->
  <arg name="controller"                         default="rmp"/>

  <!-- Robot specs  -->
  <arg name="robot_length"     default="0.6" />
  <arg name="robot_width"      default="0.4" />
  <arg name="robot_height"     default="0.4" />
  <arg name="robot_clearance"  default="0.1" />

  <!-- motion_mode: 0: turnwalkturn or 1: shuffle -->
  <arg name="motion_mode"                        default="0"/>
  <arg name="goal_distance_threshold"            default="0.15"/>
  <arg name="pose_input_topic"                   default="/state_estimator/pose_in_odom"/>
  <arg name="output_topic"                       default="/mobile_base_controller/cmd_vel"/>
  <arg name="elevation_map_topic"                default="/elevation_mapping_lidar/elevation_map_raw"/>

  <arg name="output_type"                        default="twist_stamped"/>
  <arg name="fixed_frame"                        default="odom"/>
  <arg name="base_frame"                         default="base_link"/>
  <arg name="goal_behind_mode"                   default="1"/>
  <!-- Only goals received in the given frames will be accepted  -->
  <arg name="valid_goal_frames"                  default="[map, odom, teach_map]"/>

  <arg name="max_forward_linear_velocity"        default="0.3" />
  <!-- Extrm cannot move laterally, but this cannot be set to zero. Diff drive
  arg zeroes this anyway -->
  <arg name="max_lateral_linear_velocity"        default="0.05" />
  <arg name="max_angular_velocity"               default="0.35"/>
  <!-- Can't set this to zero either -->
  <arg name="max_turning_linear_velocity"        default="0.01" />
  <!-- If the control command is nonzero and less in absolute value than the
  minimum value, it will be set to the minimum value. -->
  <arg name="min_angular_velocity"               default="0.2"/>
  <arg name="min_linear_velocity"                default="0.15"/>
  <!-- If this value is non-zero, control commands will not exceed this rate.
  The rate of the control commands is defined by the rate at which messages
  are received on the pose input topic. -->
  <arg name="maximum_control_command_rate_hz"    default="10"/>
  <!-- If set, the controller will not publish twist values on the y axis.
  This is a naive method and simply sets y values computed by the controller
  to zero. -->
  <arg name="differential_drive"                 default="true"/>
  <!-- If set to true, when the controller wants to turn only yaw velocities
  will be applied. The default is to apply both linear and angular velocities
  when turning. This is likely to improve performance on wheeled robots -->
  <arg name="yaw_exclusive_turns"                default="true"/>
  <!-- Ignore the heading of the goal passed to the position controller. When
  the robot reaches the goal it will not rotate. This only applies to goals which are-->
  <arg name="ignore_intermediate_goal_heading"   default="true"/>
  
  <include file="$(find locally_reactive_controller)/launch/controller.launch">
    <arg name="controller" value="$(arg controller)"/>
    <arg name="robot_length"     value ="$(arg robot_length)" />
    <arg name="robot_width"      value ="$(arg robot_width)" />
    <arg name="robot_height"     value ="$(arg robot_height)" />
    <arg name="robot_clearance"  value ="$(arg robot_clearance)" />
    
    <arg name="elevation_map_topic" value="$(arg elevation_map_topic)"/>
    <arg name="motion_mode" value="$(arg motion_mode)"/>
    <arg name="goal_distance_threshold" value="$(arg goal_distance_threshold)"/>
    <arg name="pose_input_topic" value="$(arg pose_input_topic)"/>
    <arg name="output_topic" value="$(arg output_topic)"/>
    <arg name="output_type" value="twist"/>
    <arg name="fixed_frame" value="$(arg fixed_frame)"/>
    <arg name="base_frame" value="$(arg base_frame)"/>
    <arg name="goal_behind_mode" value="$(arg goal_behind_mode)"/>
    <arg name="max_forward_linear_velocity" value="$(arg max_forward_linear_velocity)" />
    <arg name="max_lateral_linear_velocity" value="$(arg max_lateral_linear_velocity)" />
    <arg name="max_angular_velocity" value="$(arg max_angular_velocity)" />
    <arg name="max_turning_linear_velocity" value="$(arg max_turning_linear_velocity)" />
    <arg name="min_angular_velocity" value="$(arg min_angular_velocity)"/>
    <arg name="min_linear_velocity" value="$(arg min_linear_velocity)"/>
    <arg name="maximum_control_command_rate_hz" value="$(arg maximum_control_command_rate_hz)"/>
    <arg name="differential_drive" value="$(arg differential_drive)"/>
    <arg name="yaw_exclusive_turns" value="$(arg yaw_exclusive_turns)"/>
    <arg name="ignore_intermediate_goal_heading" value="$(arg ignore_intermediate_goal_heading)"/>
</include>

  <include file="$(find locally_reactive_controller)/launch/throttle_collision_map.launch"/>

</launch>
